### PR TITLE
Burn slashing penalties

### PR DESCRIPTION
**Pull request by haasted**
Slashing penalties are now burned, rather than distributed as rewards.

Time-based (rather than blockcount-based) slashing and jailing is unchanged.

Components from the existing slashing module in the SDK are referenced as much as possible, rather than using forks.

### CODE REVIEW

**Comment by blewater** [comment]
I pushed a small fix for the staking test

File: x/slashing/keeper/infractions.go [hunk 1]
```diff
@@ -2,11 +2,10 @@ package keeper
...
-	"github.com/tendermint/tendermint/crypto"
-	db "github.com/tendermint/tm-db"
-
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/x/slashing/types"
+	"github.com/tendermint/tendermint/crypto"
+	db "github.com/tendermint/tm-db"
 )
 
 func (k Keeper) HandleValidatorSignature(ctx sdk.Context, batch db.Batch, addr crypto.Address, power int64, signed bool, blockCount int64, slashable bool) {
```
Comment (blewater) [inline x/slashing/keeper/infractions.go:11]: Line 25: Github does not allow inserting a comment on unmodified code, there's a Todo for emitting an EventTypeLiveness that can be copied from the Sdk.
Line 35: another Todo

File: x/slashing/keeper/keeper.go [hunk 2]
```diff
@@ -176,34 +117,6 @@ func (k Keeper) setBlockTimes(batch db.Batch, blockTimes []time.Time) {
...
-	}
-
-	activePenalties := k.getPendingPenalties()
-	activePenalties.Add(consAddr.String(), coins)
-	k.setPendingPenalties(batch, activePenalties)
-	k.Logger(ctx).Info("set penalties", "active", activePenalties)
-}
-
 // Adopted from cosmos-sdk/x/staking/keeper/slash.go
 func calculateSlashingAmount(power int64, slashFactor sdk.Dec) sdk.Int {
```
Comment (blewater) [inline x/slashing/keeper/keeper.go:121]: Does not appear to be used anymore after the removal of slashValidator.

**Review by blewater — CHANGES_REQUESTED** [review]
keeper_test also seems to need to be looked up, lots of commented assertions that appear true in this or another issue.

File: x/slashing/keeper/infractions.go (hunk 1 above)
Comment (haasted) [reply x/slashing/keeper/infractions.go:11]: Added Liveness event and removed TODOs.

File: x/slashing/keeper/keeper.go (hunk 2 above)
Comment (haasted) [reply x/slashing/keeper/keeper.go:121]: Removed.

File: x/slashing/keeper/keeper.go [hunk 3]
```diff
@@ -37,9 +36,9 @@ type Keeper struct {
...
 	key sdk.StoreKey,
-	sk types.StakingKeeper,
-	paramspace types.ParamSubspace,
-	bankKeeper types.BankKeeper,
+	sk sdkslashingtypes.StakingKeeper,
+	paramspace sdkslashingtypes.ParamSubspace,
+	bankKeeper sdkslashingtypes.BankKeeper,
 	database types.ReadOnlyDB,
 	feeModuleName string,
 ) Keeper {
```
[the remark this answers is not in the document]
Comment (haasted) [reply x/slashing/keeper/keeper.go:44]: I'm ok with using the `gob` package here. It's not externally facing, and replacing it with pb doesn't provide a lot of value at this point.

[the remark this answers is not in the document]
Comment (blewater) [reply x/slashing/keeper/keeper.go:44]: Gob is not designed with determinism in mind. The community has identified a couple of corner cases i.e., as this simplified https://play.golang.org/p/lZF5ShjFh_h

File: networktest/eventlistener.go [hunk 4]
```diff
@@ -81,18 +81,6 @@ func (el EventListener) awaitQuery(query string) (func() *ct.ResultEvent, error)
...
-	}
-
-	return func() bool {
-		evt := eventFn()
-		return evt != nil
-	}, nil
-}
-
 func (el EventListener) AwaitSlash() (func() *abcitypes.Event, error) {
 	// todo (reviewer): please note the new query format. See https://docs.cosmos.network/master/core/events.html#subscribing-to-events
```
Comment (blewater) [inline networktest/eventlistener.go:85]: Noted at this point and this can now be removed.

Comment (blewater) [inline networktest/eventlistener.go:85]: Noted at this point and this can now be removed.

File: x/slashing/keeper/keeper.go [hunk 5]
```diff
@@ -94,64 +93,6 @@ func (k Keeper) deleteMissingBlocksForValidator(batch db.Batch, address sdk.Cons
 	batch.Delete(key)
```
Comment (blewater) [inline x/slashing/keeper/keeper.go:93]: if this errs, could it create problems for the validator? Maybe some logging?
Also, I noticed that some erred keeper operations panic but this is invisible. i.e., 
https://github.com/e-money/em-ledger/blob/b893cdc2432e711264eae77358309f9934bc64ec/x/slashing/keeper/keeper.go#L73
https://github.com/e-money/em-ledger/blob/b893cdc2432e711264eae77358309f9934bc64ec/x/slashing/keeper/keeper.go#L84
https://github.com/e-money/em-ledger/blob/b893cdc2432e711264eae77358309f9934bc64ec/x/slashing/keeper/keeper.go#L99

**Review by blewater — COMMENTED** [review]
Sorry I had not noticed a few things before, i.e. a few Todo items that have been resolved and can be removed (Gob) or could be looked at.
https://github.com/e-money/em-ledger/blob/b893cdc2432e711264eae77358309f9934bc64ec/x/slashing/keeper/keeper.go#L60
https://github.com/e-money/em-ledger/blob/b893cdc2432e711264eae77358309f9934bc64ec/x/slashing/keeper/keeper.go#L99

File: x/slashing/keeper/keeper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (haasted) [reply x/slashing/keeper/keeper.go:44]: Leaving a note of our previous conversation regarding this here: The Playground example is rather unnerving. Must be some global state somewhere in the `gob` module. That being said, the stuff encoded using `gob` is not added to the app_state, so it provides very little real-world value to refactor this to another serialization format. We'll leave it with `gob` for now.

File: x/slashing/keeper/keeper.go (hunk 5 above)
Comment (haasted) [reply x/slashing/keeper/keeper.go:93]: If the errors above occur, they'll very shortly after make the validator lose consensus from the network. Panicking at these places ensure that there is at least a signal about the root cause location.

**Comment by haasted** [comment]
I've removed the `// TODO Better handling` comments. Could not come up with an actual, better way to handle the issues. :)

**Review by blewater — APPROVED** [review]
Great nuanced work!

One observation to note, the unit test 
`keeper_test/TestValidatorDippingInAndOut` midway returns a zero-length slice from `truncateByWindow()` because the binary search at this line
https://github.com/e-money/em-ledger/blob/9a8547412eac7cd4a8db7804b1b63df5c49cf890/x/slashing/keeper/abci.go#L45
can't find a time element beyond the threshold.
Go allows a subrange starting with an 'index == len(slice)` without crashing but can't tell the biz logic implications of empty time slices by the truncating function.

**Pull request merged**